packages/upload-api/src/ucan/conclude.js

```diff
@@ -1,106 +1,33 @@
-        return {
-          error: new ReferencedInvocationNotFound(ranInvocation.link()),
-        }
-      }
-      return httpPutTaskGetRes
-    }
-
-    // Store receipt
-    const receiptPutRes = await receiptsStorage.put(receipt)
-    if (receiptPutRes.error) {
-      return {
-        error: receiptPutRes.error,
-      }
-    }
-
-    // THIS IS A TEMPORARY HACK
-    // Schedule `blob/accept` if there is a `http/put` capabilities
-    // inside the invocation that this receipt comes from
-    const scheduleRes = await Promise.all(
-      httpPutTaskGetRes.ok.capabilities
-        // Go through invocation tasks and get all `http/put`
-        .filter((cap) => cap.can === HTTP.put.can)
-        // @ts-expect-error body exists in `http/put` but unknown type here
-        .map(async (/** @type {API.HTTPPut} */ cap) => {
-          // Get triggering task (blob/allocate) by checking blocking task from `url`
-          /** @type {API.UnknownLink} */
-          // @ts-expect-error ts does not know how to get this
-          const [, blobAllocateTaskCid] = cap.nb.url['ucan/await']
-          const blobAllocateTaskGet = await tasksStorage.get(
-            blobAllocateTaskCid
-          )
-          if (blobAllocateTaskGet.error) {
-            return blobAllocateTaskGet
-          }
-
-          /** @type {API.BlobAllocate} */
-          // @ts-expect-error ts does not know how to get this
-          const allocateCapability = blobAllocateTaskGet.ok.capabilities.find(
-            // @ts-expect-error ts does not know how to get this
-            (/** @type {API.BlobAllocate} */ allocateCap) =>
-              equals(allocateCap.nb.blob.digest, cap.nb.body.digest) &&
-              allocateCap.can === W3sBlob.allocate.can
-          )
-
-          const blobAccept = await W3sBlob.accept
-            .invoke({
-              issuer: id,
-              audience: id,
-              with: id.did(),
-              nb: {
-                blob: cap.nb.body,
-                space: allocateCapability.nb.space,
-                _put: {
-                  'ucan/await': ['.out.ok', ranInvocation.link()],
-                },
-              },
-              // Expiry is set to `Infinity` so that CID will come out the same
-              // as returned in effect on `blob/add`
-              expiration: Infinity,
-            })
-            .delegate()
-
-          return tasksScheduler.schedule(blobAccept)
-        })
-    )
-
-    const scheduleErrors = scheduleRes.filter((res) => res.error)
-    if (scheduleErrors.length && scheduleErrors[0].error) {
-      return {
-        error: scheduleErrors[0].error,
-      }
+    // First of all that we create a message and save it to agent store, that
+    // way even if invocation occurs later we will have receipt already.
+    const message = await Message.build({ receipts: [receipt] })
+    const save = await context.agentStore.messages.write(message)
+    if (save.error) {
```

Comment:
also, I don't think we should save a message with a receipt for an unknown invocation (which is prevented in previous code)

Reply:
@vasco-santos I attempted to provide some rational behind the change in the comments above. Please let me if you're ok with storing receipts without having to find invocations for those or if you do still think they should be rejected.

Reply:
I removed this lines entirely since I have since realized that all of this was kind of a moot point. If we received message with `ucan/conclude` invocation we have already stored receipt contained by it either way. So we really don't need to create another message and save it again regardless whether we are aware of invocation or not.

What I end up doing is made `AgentStore` aware of `ucan/conclude` so it would index contained receipt.

In the end things end up looking even simpler because ucan conclude handler doesn't really need to do anything other than poll the accept handler if receipt is for the `http/put`.

Reply:
> I removed this lines entirely since I have since realized that all of this was kind of a moot point. If we received message with ucan/conclude invocation we have already stored receipt contained by it either way. So we really don't need to create another message and save it again regardless whether we are aware of invocation or not.

That seems reasonable if we make sure AgentStore implementations really make the indexing